### DISCUSSION TITLE
ExportAsCsv on Laravel Vapor downloads unknown.txt

### DISCUSSION THREAD

**Discussion by dmason30** [laravel/nova-issues#4886] [Ideas & Feature Requests] [answered]
- Laravel Version: 9
- Nova Version: 4.2.5
- PHP Version: 8.1.3

### Description:
When running the `ExportAsCsv` action on vapor it ends up with downloading `unknown.txt` with contents `false` see video, whereas locally it works fine:

https://user-images.githubusercontent.com/20278756/165036127-805b66d7-55c7-4c32-b111-8e8e14297c2f.mov

### Possible Solution:
Does the vapor binary header need to be added to the download response?

```php
'X-Vapor-Base64-Encode' => 'True',
```

**Comment by crynobone** [comment]
Can you test and verify if the following works

```php
<?php

namespace App\Nova\Actions; 

use Laravel\Nova\Actions\ExportAsCsv as NovaExportAsCsv;
use Laravel\Nova\Http\Requests\ActionRequest;

class ExportAsCsv extends NovaExportAsCsv
{
    /**
     * Execute the action for the given request.
     *
     * @param  \Laravel\Nova\Http\Requests\ActionRequest  $request
     * @return mixed
     *
     * @throws \Laravel\Nova\Exceptions\MissingActionHandlerException|\Throwable
     */
    public function handleRequest(ActionRequest $request)
    {
        return tap(parent::handleRequest($request), function ($response) {
            $response->headers->set('X-Vapor-Base64-Encode', 'True');
        });
    }
}
```

And then use `App\Nova\Actions\ExportAsCsv` instead of `Laravel\Nova\Actions\ExportAsCsv`

**Comment by dmason30** [comment]
@crynobone Unfortunately, I still get the `unknown.txt` but the difference is that the contents of the file and the request response are now blank rather than containing `false`.

**Comment by crynobone** [comment]
Can anyone tests this `VaporExportAsCsv` and see if it does work on Vapor.

https://gist.github.com/crynobone/56987f4ca5be4483104f692ec08d6dcb

**Comment by dmason30** [comment]
@crynobone 

### Local Environment
These may apply to ExportAsCsv too:

1. `VaporExportAsCsv::make()` 
Uses whatever default disk in config/filesytems.php is which is 'local' on new laravel applications.
The 'local' disk is unable to provide a valid download url so it creates the csv but the download fails.
Result: ⚠️ Probably just needs covering in docs

2. `VaporExportAsCsv::make(null, 'public')` 
Requires you to run `php artisan storage:link` and then the download url works correctly
Result: ✅  Success

3. `VaporExportAsCsv::make(null, 's3')` 
Produces the below direct S3 download url which failed locally I assume as it requires some configuration in AWS to make it publicly accessible.
https://xyz-dev-bucket.s3.eu-west-1.amazonaws.com/nova-actions-export-as-csv/cEZwKR081qVIodEJ8YBKuoItKpSMYFojrofnvqak.csv
Result: ✅ Assume success but docs may need to provide the required AWS S3 config.

### Vapor Environment

1. `VaporExportAsCsv::make(null, 's3'):`
Caused exception: https://flareapp.io/share/17DMxdR5#F99
fopen(export-as-csv-20220609121407.csv): Failed to open stream: Read-only file system
Result: ❌ Failed

<details>
<summary>Click here for raw log</summary>
<code>
{message: "fopen(export-as-csv-20220609121407.csv): Failed to open stream: Read-only file system",…}
exception: "ErrorException"
file: "/var/task/vendor/box/spout/src/Spout/Common/Helper/GlobalFunctionsHelper.php"
line: 23
message: "fopen(export-as-csv-20220609121407.csv): Failed to open stream: Read-only file system"
trace: [,…]
0: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php",…}
1: {function: "Illuminate\Foundation\Bootstrap\{closure}",…}
2: {file: "/var/task/vendor/box/spout/src/Spout/Common/Helper/GlobalFunctionsHelper.php", line: 23,…}
3: {file: "/var/task/vendor/box/spout/src/Spout/Writer/WriterAbstract.php", line: 102, function: "fopen",…}
4: {file: "/var/task/vendor/rap2hpoutre/fast-excel/src/Exportable.php", line: 99, function: "openToFile",…}
5: {file: "/var/task/vendor/rap2hpoutre/fast-excel/src/Exportable.php", line: 48,…}
6: {file: "/var/task/app/Nova/Actions/VaporExportAsCsv.php", line: 93, function: "export",…}
7: {file: "/var/task/app/Nova/Actions/VaporExportAsCsv.php", line: 80, function: "dispatchExportUsing",…}
8: {file: "/var/task/vendor/laravel/nova/src/Actions/Action.php", line: 294,…}
9: {file: "/var/task/vendor/laravel/nova/src/Actions/DispatchAction.php", line: 197,…}
10: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Support/helpers.php", line: 418,…}
11: {file: "/var/task/vendor/laravel/nova/src/Actions/DispatchAction.php", line: 116, function: "with"}
12: {file: "/var/task/vendor/laravel/nova/src/Actions/Action.php", line: 308, function: "dispatch",…}
13: {file: "/var/task/vendor/laravel/nova/src/Http/Controllers/ActionController.php", line: 46,…}
14: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Controller.php", line: 54,…}
15: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php", line: 45,…}
16: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Route.php", line: 261,…}
17: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Route.php", line: 204,…}
18: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Router.php", line: 725,…}
19: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 141,…}
20: {file: "/var/task/vendor/laravel/nova/src/Http/Middleware/Authorize.php", line: 18,…}
21: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
22: {file: "/var/task/vendor/laravel/nova/src/Http/Middleware/BootTools.php", line: 20,…}
23: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
24: {file: "/var/task/vendor/laravel/nova/src/Http/Middleware/DispatchServingNovaEvent.php", line: 24,…}
25: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
26: {file: "/var/task/vendor/inertiajs/inertia-laravel/src/Middleware.php", line: 88,…}
27: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
28: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php",…}
29: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
30: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Auth/Middleware/Authenticate.php", line: 44,…}
31: {file: "/var/task/vendor/laravel/nova/src/Http/Middleware/Authenticate.php", line: 31,…}
32: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
33: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Auth/Middleware/Authenticate.php", line: 44,…}
34: {file: "/var/task/app/Http/Middleware/Authenticate.php", line: 14, function: "handle",…}
35: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
36: {,…}
37: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
38: {,…}
39: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
40: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php",…}
41: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php",…}
42: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
43: {,…}
44: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
45: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php",…}
46: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
47: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 116,…}
48: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Router.php", line: 726,…}
49: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Router.php", line: 703,…}
50: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Router.php", line: 667,…}
51: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Routing/Router.php", line: 656,…}
52: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php", line: 167,…}
53: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 141,…}
54: {file: "/var/task/vendor/livewire/livewire/src/DisableBrowserCache.php", line: 19,…}
55: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
56: {file: "/var/task/vendor/laravel/nova/src/Http/Middleware/ServeNova.php", line: 23,…}
57: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
58: {file: "/var/task/vendor/laravel/vapor-core/src/Http/Middleware/ServeStaticAssets.php", line: 21,…}
59: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
60: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Http/Middleware/HandleCors.php", line: 49,…}
61: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
62: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php", line: 39,…}
63: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
64: {,…}
65: {,…}
66: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
67: {,…}
68: {,…}
69: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
70: {,…}
71: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
72: {,…}
73: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
74: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 116,…}
75: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php", line: 142,…}
76: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php", line: 111,…}
77: {file: "/var/task/vendor/laravel/octane/src/ApplicationGateway.php", line: 36, function: "handle",…}
78: {file: "/var/task/vendor/laravel/octane/src/Worker.php", line: 92, function: "handle",…}
79: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/Octane/Octane.php", line: 190,…}
80: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 141,…}
81: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/Http/Middleware/EnsureBinaryEncoding.php",…}
82: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
83: {,…}
84: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
85: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/Http/Middleware/RedirectStaticAssets.php",…}
86: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
87: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/Http/Middleware/EnsureOnNakedDomain.php",…}
88: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 180,…}
89: {file: "/var/task/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php", line: 116,…}
90: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/Octane/Octane.php", line: 193,…}
91: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/Octane/Octane.php", line: 158,…}
92: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/Handlers/OctaneHandler.php", line: 26,…}
93: {file: "/var/task/octaneRuntime.php", line: 87, function: "handle",…}
94: {file: "/var/task/vendor/laravel/vapor-core/src/Runtime/LambdaRuntime.php", line: 53,…}
95: {file: "/var/task/octaneRuntime.php", line: 89, function: "nextInvocation",…}
96: {file: "/var/task/runtime.php", line: 33, function: "require"}
97: {file: "/opt/bootstrap.php", line: 6, function: "require"}
</code>
</details>

**Comment by dmason30** [comment]
Here are results using the changes recommended here:
https://gist.github.com/crynobone/56987f4ca5be4483104f692ec08d6dcb?permalink_comment_id=4194441#gistcomment-4194441

### Local Environment
These may apply to ExportAsCsv too:

1. `VaporExportAsCsv::make('Export As CSV Local')` 
Result: ❌ Local disk probs should not need to work

2. `VaporExportAsCsv::make('Export As CSV Public', 'public')` 
Result: ✅  Success

3. `VaporExportAsCsv::make('Export As CSV Cloud', 's3')` 
Result: ✅ Success

### Vapor Environment
1. `VaporExportAsCsv::make('Export As CSV Cloud', 's3')` 
25 rows result: ✅ Success 
90000 rows result: ⚠️ Endpoint request timed out 
(API Gateway only allows a request to run for 30 seconds).

2. `VaporExportAsCsv::make('Export As CSV Cloud', 's3')->withTypeSelector()`
XLSX file type selected
25 rows result: ✅ Success 
### Overall ✅ Success 🎉 🥳

Couple ideas for features:
1. Provide a `withStorageDisk('s3')` method for nicer DX when we don't want to customise action name?
**Before** `VaporExportAsCsv::make(null, 's3')`
**After** `VaporExportAsCsv::make()->withStorageDisk('s3')`

2. Consider adding a queueable version of the exports for larger datasets (especially on vapor)? 
Download url could be sent as a NovaNotification or for more flexibility dispatch an Event with the User and download url so that developers can register listeners as need.

**Comment by crynobone** [comment]
Queueable can't be done without serializing ActionRequest and query builder and this is somewhat hackish (which is being done on <https://github.com/SpartnerNL/Laravel-Nova-Excel> using <https://github.com/laravie/serialize-queries>).

**Comment by crynobone** [comment, marked as answer]
@dmason30 

I just released the following tools which should solve the issues on Vapor:

* https://github.com/nova-kit/nova-queued-export-as-csv 
* https://github.com/nova-kit/nova-on-vapor
